### DISCUSSION TITLE
Entity-Component-System paradigm

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#17584] [Feature requests]
ECS would solve the issues relating to lack of room in track element. I am not very familiar with ECS paradigms, so I would not be the perfect integrator. That said, I want this yesterday, so anything that can get this feature moving is something I am all in on.

Here's some questions and assumptions I have about ECS as it relates to OpenRCT2:
- Each TileElement Entity will refer to only one Component?
- Should Components be stored in different iterables by type or a single iterables for all types? If the latter, how will this affect struct size constraints?
- How does System work? Are object methods just gone, reverted back to pure C pass-the-pointer style? Components can have object methods right?
- Would Entities live in the same place TileEntities do currently, where the XY position is inferred?

I have wanted extra TrackElement data for a long time, so I am all in on this.

**Comment by Broxzier** [comment]
Moving to ECS is no small task, especially if we want to keep everything working while moving towards such a system.

> Each TileElement Entity will refer to only one Component?

No, it's quite the opposite. Each element is an entity, and this can have a flexible amount of components.

It could even be so that tiles themselves are entities that only have a tile location component (and perhaps a 'visible' component for edge tiles), and they have components which make up the tile elements, however this would likely be more work and can result in confusing code.

> Should Components be stored in different buckets by type or a single bucket? If the latter, how will this affect struct size constraints?

One bucket per component. This does not affect stuct size, because they are not part of the tile element structs.

> How does System work? Are object methods just gone, reverted back to pure C pass-the-pointer style? Components can have object methods right?

Depends on the implementation. I do not have a clear idea of how this is implemented exactly. Perhaps @duncanspumpkin has a more clear idea here.

> Would Entities live in the same place TileEntities do currently, where the XY position is inferred?

With 'Entities' do you refer to the ones we currently have (peeps, vehicles, balloons, etc)? Position should be a component, not some entity identifier.

What I expect to be the most difficult part is for systems to find relative neighbouring tile elements.

As as a first step, I think giving tile elements that we want to give lesser-used components an entity ID that can then refer to the components that are stored elsewhere. This keeps the more-used components in the memory blocks of the tile elements themselves, meaning no big performance drop.

Note that once tile elements do use ECS entirely, performance is expected to be much better, since related fields are packed together in memory.

**Reply by spacek531** [reply]
> No, it's quite the opposite. Each element is an entity, and this can have a flexible amount of components.

What is the benefit of multiple components per entity in OpenRCT2? How would components be broken down? You mentioned a "tile location component," is this a Component or just part of the regular struct?

>> Would Entities live in the same place TileEntities do currently, where the XY position is inferred?
>
> With 'Entities' do you refer to the ones we currently have (peeps, vehicles, balloons, etc)? Position should be a component, not some entity identifier.

This should read, "Would new Tile entities live in the same place TileElements do currently?" which is to imply separate buckets for current TileElements and current Entities' new Entities.

>> Should Components be stored in different buckets by type or a single bucket? If the latter, how will this affect struct size constraints?
>
> One bucket per component. This does not affect stuct size, because they are not part of the tile element structs.

I meant to say iterable (vector or array), not bucket.

**Comment by spacek531** [comment]
If I'm reading between the lines properly, each Entity gets an iterable of components and a type enum to tell the System what components it can expect in what order? Components should be broken into small sizes, like a single CoordsXYZ for position, or a uint16_t for Color, or a few consecutive bytes as a peep's inventory.

If I've misunderstood, would you explain with an example such as how a TrackElement would be broken down in to Components?

**Comment by IntelOrca** [comment]
The array of tile elements can already be considered partially ECS. Each tile element has IDs that refer to other things, like what object an element is using or simply just for more information. Take banner element for example, which refers to a banner ID, something that is a 1:1 relationship only so that the element has more data.

The memory structure for tile elements, the flag for when the stack ends, and the tile element grid is all designed specifically so for efficiency and small memory footprint.

True ECS would allow as many components as you like per tile element, and I don't see how that would be practical without some heavy compromises.

**Comment by duncanspumpkin** [comment]
Yes its not going to be simple to move towards this but the idea would be generally each tile element gets an entity id assigned so that you can look up the components for it. Initially that would just be a singular component that would map almost exactly the same as the current tile element (except you could easily grow it in size).

```
Existing:

[Tile]
TrackElement, (properties: height, flags, track type id, ride id, etc...)
SurfaceElement, (properties: height, flags, surface type id, etc...)



New (1st stage):

[Tile]
TrackElement, (properties: height, flags, entity id, etc..)
SurfaceElement, (properties: height, flags, entity id, etc..)

[Entity Registry]
Map between entity id and component.

[Track Component Store]
TrackComponent, (properties: track type id, ride id, etc..)

[Surface Component Store]
SurfaceComponent, (properties: surface type id, etc..)



New (final stage):

[Entity Registry]
Map between entity id and components.

[Tile Component Store]
TileComponent, (properties: multiple entity ids {each entity id might have multiple components})

[Base Element Component Store]
Element, (properties: height, flags, etc..)

[Track Component Store]
TrackComponent, (properties: track type id, ride id, etc..)

[Surface Component Store]
SurfaceComponent, (properties: surface type id, etc..)
```
Something like this but keep in mind this is just a mock up idea. Same would also be done for what we have atm called `Entities` (peeps, vehicles) they would also become components but they need lots of breaking up. We could reuse the current entity registry code for all of this but its super inefficient as its all based on linked lists. I was sort of hoping of solving that first then bringing in tile elements into this.

The main benefits of all of this is that you can much more efficiently iterate things and save things tightly. For example we waste a lot of time computing the exact same rendering data for a tile that could be saved in a component to vastly speed up drawing.

We wouldn't really be changing much in terms of functions and classes apart from continuing to split them up more.

**Comment by spacek531** [comment]
Is it agreeable that replacing the Banner's component store with a generic component store that can be applied to any class is a good first step for ECS implementation?

**Reply by spacek531** [reply]
I've begun a branch to start implementing this. I haven't made it a PR because I have enough of those already. Here's what I have, if there's any issues or improvements please let me know: https://github.com/spacek531/OpenRCT2/tree/component-store
